### PR TITLE
AMQP-640: DirectMLContainer - Docs and XML Parser

### PR DESCRIPTION
**Pull request by garyrussell**
JIRA: https://jira.spring.io/browse/AMQP-640

### CODE REVIEW

File: spring-rabbit/src/main/java/org/springframework/amqp/rabbit/listener/AbstractMessageListenerContainer.java [hunk 1]
```diff
@@ -1481,4 +1500,31 @@ protected WrappedTransactionException(Throwable cause) {
...
+					if (logger.isWarnEnabled()) {
+						logger.warn(message + ": " + cause.toString());
+					}
+				}
+				else if (!RabbitUtils.isNormalChannelClose(cause)) {
+					logger.error(message + ": " + cause.getMessage());
+				}
+			}
+			else {
+				logger.error("Unexpected invocation of " + this.getClass() + ", with message: " + message, t);
```
Comment (artembilan) [inline spring-rabbit/src/main/java/org/springframework/amqp/rabbit/listener/AbstractMessageListenerContainer.java]: Why don't we use `is...Enable()` patter for `error()`? How about if we have logger config just for `FATAL`?
(Yes, this is existing code, but still...)

File: src/reference/asciidoc/amqp.adoc [hunk 2]
```diff
@@ -1283,8 +1283,12 @@ Basically, the container handles the "active" responsibilities so that the liste
 The container is an example of a "lifecycle" component.
 It provides methods for starting and stopping.
 When configuring the container, you are essentially bridging the gap between an AMQP Queue and the `MessageListener` instance.
-You must provide a reference to the `ConnectionFactory` and the queue name or Queue instance(s) from which that listener should consume Messages.
-Here is the most basic example using the default implementation, `SimpleMessageListenerContainer` :
+You must provide a reference to the `ConnectionFactory` and the queue name(s) or Queue instance(s) from which that listener should consume Messages.
+
+With versions prior to _version 2.0_, the was one listener container - the `SimpleMessageListenerContainer`; there is now a second container - the `DirectMessageListenerContainer`.
```
Comment (artembilan) [inline src/reference/asciidoc/amqp.adoc]: Typo: the**re** was

File: src/reference/asciidoc/amqp.adoc [hunk 3]
```diff
@@ -2084,30 +2102,60 @@ Starting with _version 1.5_, you can now assign a `group` to the container on th
...
+===== Choosing a Container
+
+_Version 2.0_ introduced the `DirectMessageListenerContainer` (DMLC); previously, only the `SimpleMessageListenerContainer` (SMLC) was available.
+The SMLC uses an internal queue and a dedicated thread for each consumer; if a container is configured to listen to multiple queues, the same consumer thread is used to process all the queues.
+Concurrency is controlled by `concurrentConsumers` and other properties.
+As messages arrive from the RabbitMQ client, the client thread hands them off to the consumer thread via the queue.
+This architecture was required because in early versions of the RabbitMQ client, multiple concurrent deliveries were not possible.
+Newer versions of the client have a revised threading model and can now support concurrency.
+This has allowed the introduction of the DMLC where the listener is now invoked directly on the RabbitMQ Client thread.
+It's architecture is therefore actually "simpler" than the SMLC.
```
Comment (artembilan) [inline src/reference/asciidoc/amqp.adoc]: Typo:  Its architecture is

File: src/reference/asciidoc/amqp.adoc [hunk 4]
```diff
@@ -3789,119 +3853,189 @@ if its enclosing `ApplicationContext` is closed) it waits for in-flight messages
...
 (error-handler)
 
 | A reference to an ErrorHandler strategy for handling any uncaught Exceptions that may occur during the execution of the MessageListener.
 Default: `ConditionalRejectingErrorHandler`
 
+a| image::images/tickmark.png[]
+a| image::images/tickmark.png[]
+
+| consumersPerQueue
+(consumers-per-queue)
```
Comment (artembilan) [inline src/reference/asciidoc/amqp.adoc]: I see `DMLC` `monitorInterval` and `taskScheduler` properties are missed in this table.
WDYT?

File: spring-rabbit/src/main/java/org/springframework/amqp/rabbit/listener/AbstractMessageListenerContainer.java (hunk 1 above)
Comment (garyrussell) [inline spring-rabbit/src/main/java/org/springframework/amqp/rabbit/listener/AbstractMessageListenerContainer.java]: I can't get too excited about this - it should never get called so I don't really think the overhead will ever be invoked.

**Comment by artembilan** [comment]
Merged as https://github.com/spring-projects/spring-amqp/commit/d955f7d3909d7a09c43e584e45213f39f843de32

**Pull request closed without merging**